Dockerfile

```diff
@@ -32,12 +32,17 @@ RUN set -ex && \
             git \
             tzdata \
             zlib1g-dev \
+            curl \
             $(cat /tmp/ruby_build_deps.txt) && \
             apt-get clean && rm -r /var/lib/apt/lists/*
 
 RUN set -ex && \
     useradd -ms /bin/bash ubuntu
 
+RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
```

Comment:
Two things:

  - I think YJIT doesn't require a particularly recent Rust, so we might be able to install the relevant debian package instead of using rust up.
  - Also rustup is like 1.5GiB, so this will blow up the image size significantly. It would be best to remove it once Ruby is installed like it's done for `/tmp/ruby_build_deps.txt`

Reply:
We intentionally stuck to Rust >= 1.58.1 to make it easier to build YJIT with prepackaged Rust installs 👍

Reply:
I took this line directly from ruby/ruby here https://github.com/ruby/ruby/blob/e896b338605914df70c1c34274f5147dff6f71a3/.cirrus.yml#L110. From @maximecb's comment I can't tell if you're agreeing I should change it or that it should stay as-is.

Reply:
I don't remember why I used rustup for cirrus CI, I'll have to look at whether the rustc from the repositories is recent enough, but it's not a big deal there because it's ephemeral. 

And yeah @maximecb is agreeing that it should be changed.

Reply:
I have not been able to get a working version of this with the apt 1.59.0 version of Rust. It refuses to build Ruby with when running `rake docker:build ruby_version=master:b3d8dddee7a9ea0bc9c278a5c9faa4df81afd57e` or `rake docker:build arch=arm64 ruby_version=master:b3d8dddee7a9ea0bc9c278a5c9faa4df81afd57e`. I get the following error when compiling.

```
#13 126.1 error[E0433]: failed to resolve: could not find `is_aarch64_feature_detected` in `arch`
#13 126.1    --> /usr/src/ruby/yjit/src/options.rs:157:28
#13 126.1     |
#13 126.1 157 |             if !std::arch::is_aarch64_feature_detected!("lse") {
#13 126.1     |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^ could not find `is_aarch64_feature_detected` in `arch`
#13 126.1 
```

It seems like something installed the wrong architecture but I don't understand why this worked fine with `rustup`. For reference I run my M1 terminal in rosetta so I don't have to switch back and forth.

If I set `DOCKER_DEFAULT_PLATFORM=linux/amd64` with `rake docker:build ruby_version=master:b3d8dddee7a9ea0bc9c278a5c9faa4df81afd57e` it will build but I when I try to run Ruby like `ruby --enable-yjit -v` inside the container I get a segv. I also tried with arch64 and had the same result:

```
ruby: [BUG] Couldn't make JIT page (0x0000ffff7789b000, 0 bytes) executable, errno: Cannot allocate memory
```

Reply:
I've made the change to remove Rust when we're done building Ruby but am still installing Rust via rustup.

If it's a big dealbreaker to install via rustup I can keep looking into why my machine won't build but right now I'm at a loss as to what's going on and would like some help to get it working. 

dditionally the build with 1.59 via apt-get when it does compile (but segv's) is extremely slow (20+ min) versus building with the rustup version that gets installed.

Reply:
I think this method of std::arch is probably not available on Rust 1.59... We should use Rust 1.58.1 in the YJIT CI 🤔 

I'll try to push a commit to make it so this code doesn't run in release mode 🤔

Reply:
@byroot since rustup seems to work for the moment and Eileen was able to uninstall it after, I would be willing to merge this PR as is, if you are too.

Reply:
I'm +1 to use rustc provided by `apt`. YJIT doesn't support rustc-1.59.0?

Reply:
I am not able to get a working build (see previous comments with errors) on rustc 1.59.0 provided by `apt`.

Reply:
In general I'm for this PR.

About the implementation, even though you uninstall rust as `rustup self uninstall -y`, once you finish `RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y`, it will consume 1.5GB.
It's a sort of docker image size optimization.

Reply:
> even though you uninstall rust as rustup self uninstall -y, once you finish RUN

Yes, deleting in another layer won't save space, unless the layers are squashed. Since that pattern was already done for `ruby_build_deps.txt` I assumed the layers were squashed.

Reply:
Ok I'm at a loss then, I really don't know what to do since there's no CI here running to tell me whether the apt version works and I can't get it to build correctly (without compilation failure OR segv) locally on my mac via docker.

Reply:
@eileencodes Maxime said that YJIT not compiling with 1.59 is a regression that she'll fix. 

We should make sure ARM CI use the minimum rustc version we wish to support to avoid a similar regression in the future.

Reply:
Right so I tried testing that but after her changes were merged I was unable to build anything (even without my yjit changes).